File: cargo_app/lib/utils/distance.dart
import 'dart:math';

/// Вычисляет расстояние между двумя GPS-точками по формуле гаверсинусов.
/// Возвращает расстояние в километрах.
double haversineDistance(double lat1, double lon1, double lat2, double lon2) {
  const double earthRadiusKm = 6371.0;

  final double dLat = _toRadians(lat2 - lat1);
  final double dLon = _toRadians(lon2 - lon1);

  final double a = sin(dLat / 2) * sin(dLat / 2) +
      cos(_toRadians(lat1)) *
          cos(_toRadians(lat2)) *
          sin(dLon / 2) *
          sin(dLon / 2);

  final double c = 2 * atan2(sqrt(a), sqrt(1 - a));

  return earthRadiusKm * c;
}

/// Вычисляет суммарное расстояние по массиву координат в километрах.
double calculateTotalDistance(List<Map<String, double>> track) {
  if (track.length < 2) return 0.0;

  double total = 0.0;
  for (int i = 1; i < track.length; i++) {
    total += haversineDistance(
      track[i - 1]['latitude']!,
      track[i - 1]['longitude']!,
      track[i]['latitude']!,
      track[i]['longitude']!,
    );
  }

  return (total * 10).roundToDouble() / 10; // округляем до 0.1 км
}

double _toRadians(double degrees) => degrees * pi / 180.0;

File: functions/src/distance.ts
import {GeoPoint} from "./types";

const EARTH_RADIUS_KM = 6371;

function toRadians(degrees: number): number {
  return (degrees * Math.PI) / 180;
}

/**
 * Вычисляет расстояние между двумя точками (широта/долгота)
 * по формуле гаверсинусов. Возвращает километры.
 */
export function haversineDistance(
  lat1: number,
  lon1: number,
  lat2: number,
  lon2: number
): number {
  const dLat = toRadians(lat2 - lat1);
  const dLon = toRadians(lon2 - lon1);

  const a =
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos(toRadians(lat1)) *
      Math.cos(toRadians(lat2)) *
      Math.sin(dLon / 2) *
      Math.sin(dLon / 2);

  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));

  return EARTH_RADIUS_KM * c;
}

/**
 * Вычисляет суммарное расстояние по массиву GPS-точек.
 * Точки должны быть отсортированы по времени.
 * Возвращает километры, округлённые до 1 знака после запятой.
 */
export function calculateTotalDistance(track: GeoPoint[]): number {
  if (!track || track.length < 2) {
    return 0;
  }

  let totalDistance = 0;

  for (let i = 1; i < track.length; i++) {
    const prev = track[i - 1];
    const curr = track[i];

    totalDistance += haversineDistance(
      prev.latitude,
      prev.longitude,
      curr.latitude,
      curr.longitude
    );
  }

  return Math.round(totalDistance * 10) / 10;
}
